### ISSUE TITLE
Provide guidelines around dat and time localization

### ISSUE THREAD

**Issue by stevegrossi** [lessonly/lessonly.github.io#118]
From [this thread in Slack](https://lessonly.slack.com/archives/C047M50C0/p1553266188039200?thread_ts=1553005256.006700&cid=C047M50C0):

> To my knowledge, we generally aim to display dates and times localized. Since we don't store users' time zones anywhere, that can only be done client-side with the `local_time` gem/helper or something custom. Two exceptions I'm aware of are *due dates* and *user activity*. Due dates (which don't have an explicit time component but dates still have an implicit time zone) have always been displayed as the date in Eastern Time, mainly because it's the application's default time zone (which I think should change to UTC). User Activity dates are recorded as the date in UTC because money is involved and we wanted a standard definition of a date for customers with users across the globe.

**Comment by unixmonkey** [comment]
Events specifically displays both the time and the time zone, but that's because typically an event has an actual place (and time zone).

I think that it would be helpful to add time zone to a user's profile, though we do get around it in a few places with the JS snippet `Intl.DateTimeFormat().resolvedOptions().timeZone`, but that doesn't help in mailers or anything automated.

IMO, report exports should have an option to change the time zone. Most times people want info relative to their own time zone for easy mental conversion, but for distributed teams or automated processing, there should be an option to download UTC or another zone.

Localization doesn't only refer to time zone offsets. There are formatting differences that are tied to locale as well. When displaying dates and times in reports, it's often best to be technical and stick to a database-like format of `YYYY-MM-DD HH:MM:SS`, which is unambiguous and machine parseable (plug it into Excel), regardless of locale.

Otherwise, prefer to display dates in a localized format, based on their locale/language preference.
Rails has `I18n.l(date_or_time_obj, format: :named_format)`, which complements `I18n.t()`.

For example:
```ruby
I18n.l(Date.today) # => "2019-03-27"
I18n.l(Time.current) # => "Wed, 27 Mar 2019 12:08:26 -0400"
I18n.l(Time.current, format: :full_month_date_year) # => "March 27, 2019"
I18n.locale = :de
I18n.l(Date.today) # => "27.03.2019"
I18n.l(Time.current) # => "Mi, 27. Mär 2019 12:08:00 -0400"
I18n.l(Time.current, format: :full_month_date_year) # => "27. März 2019"
```

These named formats are defined in our translation YAML files under the `date.formats` and `time.formats` keys, which we keep in separate `config/locales/localization.<locale>.yml` files, although some formats are automatically provided by Rails. If a matching format isn't found in another locale, the English version is used.

Similarly, on the frontend, we use `moment.js` formatting for dates and times, and when we want those to be localized, we use localized formats provided by the `DateFormat` object like this:
```js
moment(date).format(DateFormat.LONG_AT_TIME) // 'January 31, 2019 at 1:33 PM'
moment.locale('de')
moment(date).format(DateFormat.LONG_AT_TIME) // '31. Januar 2019 um 13:33 Uhr'
```

Sometimes we may also want a date or time in a very specific format, like for passing to a date picker widget, or our backend to `Time.parse`, or a 3rd party system. Only in those cases should we use a fixed format generated with something like `DateFormat.fixed.ISO8601_LOCAL`, or formatted with `strftime()`.

I surely have more thoughts on this, but wanted to throw this out there while it was on my mind.

**Comment by stevegrossi** [comment]
> stick to a database-like format of YYYY-MM-DD HH:MM:SS

Using a consistent, machine-parseable date format in report downloads is a great suggestion, but won't that particular format be ambiguous without a time zone? Although I did a bit of quick research and testing and it sounds like Excel doesn't support datetime formats with time zones, so perhaps our best bet is `YYYY-MM-DD HH:MM:SS` with a rule to always include "(UTC)" in the column header, e.g. "Assignment Date (UTC)".

**Comment by unixmonkey** [comment]
I knew writing that up in a hurry would bite me somewhere. I meant to include the zone part. I do think we should include time zone offset in the timestamp itself (rather than the header) like `YYYY-MM-DD HH:MM:SS -0400`, even if it is a lot of duplication. I need to brush up on what Excel does and doesn't do with certain formats, but I believe either using the zone offset should be okay, or maybe the format you get by doing `Time.now.iso8601 #=> "2019-03-29T11:41:29-04:00"`

**Comment by stevegrossi** [comment]
I agree about including the zone if we can, but I think since reports are most often consumed by a spreadsheet editor we should definitely prioritize spreadsheet-parsability. I don't have Excel to test myself, but it doesn't look like Google Sheets recognizes ISO8601 as a datetime format 🤔

**Comment by unixmonkey** [comment]
I think if we really care about Excel, then we'll probably need to be exporting xlsx files, with typed columns/cells. I'm not finding anything solid about adding zone info to timestamps, other than some advice to split them out into a separate column (or maybe add to the column header, if all are adjusted), but the db-like format without zone seems to be the common recommendation.

I did find some interesting tips here for more Excel-friendly CSVs here, but it doesn't quite sit right with me, over just providing xlsx files:
http://blog.plataformatec.com.br/2009/09/exporting-data-to-csv-and-excel-in-your-rails-app/

This has some handy info about how Excel represents times internally, and how to deal with that in Axlsx (a popular Ruby library for generating Excel files):
https://stackoverflow.com/a/13963296/23915
